### PR TITLE
Cnmf yx streamline

### PR DESCRIPTION
**Pull request by ImXman**
cnmf pipeline workable in jupyter notebook. More tests needed for terminal execution.

### CODE REVIEW

File: cellarium/ml/cli.py [hunk 1]
```diff
@@ -485,6 +485,44 @@ def nmf(args: ArgsType = None) -> None:
     )
     cli(args=args)
 
+@register_model
+def nmf_consensus(args: ArgsType = None) -> None:
```
Comment (sjfleming) [inline cellarium/ml/cli.py:489]: Do we still want this command line command?

File: cellarium/ml/core/module.py [hunk 2]
```diff
@@ -416,3 +416,23 @@ def on_save_checkpoint(self, checkpoint: dict[str, Any]) -> None:
                 checkpoint["loops"]["fit_loop"]["epoch_progress"]["total"]["completed"] += 1
                 checkpoint["loops"]["fit_loop"]["epoch_progress"]["current"]["completed"] += 1
                 checkpoint["CellariumAnnDataDataModule"]["epoch"] += 1
+
+    def on_train_end(self) -> None:
```
Comment (sjfleming) [inline cellarium/ml/core/module.py:420]: I do not think this is necessary.  See the comment on the `on_end()` function in nmf.py

File: cellarium/ml/core/module.py [hunk 3]
```diff
@@ -416,3 +416,23 @@ def on_save_checkpoint(self, checkpoint: dict[str, Any]) -> None:
...
+        """
+        Calls the ``on_epoch_end`` method on the :attr:`model` attribute.
+        If the :attr:`model` attribute has ``on_epoch_end`` method defined, then
+        ``on_epoch_end`` must be called at the end of every epoch.
+        """
+        on_end = getattr(self.model, "on_end", None)
+        if callable(on_end):
+            on_end(self.trainer)
+
+    def on_predict_end(self) -> None:
```
Comment (sjfleming) [inline cellarium/ml/core/module.py:430]: I am not sure if this is necessary.  See the comment on the `on_prediction_end()` function in nmf.py

File: cellarium/ml/models/nmf.py [hunk 4]
```diff
@@ -189,79 +477,83 @@ def solve_alpha_wKL(self, alpha_nk: torch.Tensor, x_ng: torch.Tensor, n_iteratio
...
+                updated_factors_kg.view(-1, self.n_vars) - D_buffer.view(-1, self.n_vars)
+            ) / torch.linalg.norm(updated_factors_kg.view(-1, self.n_vars))
             if D_diff <= self._D_tol:
                 break
             D_buffer = updated_factors_kg.clone()
 
         return updated_factors_kg
 
-    def forward(self, x_ng: torch.Tensor, var_names_g: np.ndarray) -> dict[str, torch.Tensor | None]:
+    def nmf_forward(self, x_ng: torch.Tensor, var_names_g: np.ndarray) -> dict[str, torch.Tensor | None]:
```
Comment (sjfleming) [inline cellarium/ml/models/nmf.py:556]: I do not think we should split things into `nmf_forward()` and `consensus_forward()`.  Can't we have just one `forward()` function?  If `r == 1`, then we are doing regular NMF, and if `r > 1` then we are doing consensus NMF.  But the code could be the exact same.  Just the `r` dimension of tensors will have size 1 if we are doing regular NMF.  Right?

File: cellarium/ml/models/nmf.py [hunk 5]
```diff
@@ -275,21 +567,67 @@ def forward(self, x_ng: torch.Tensor, var_names_g: np.ndarray) -> dict[str, torc
...
+                D_rkg = getattr(self, f"D_{i}_rkg")
+                D_rkg = self.online_dictionary_learning(x_ng=x_, factors_kg=D_rkg)
+                setattr(self, f"D_{i}_rkg", D_rkg)
 
         else:
             raise ValueError(f"Unknown algorithm: {self.algorithm}")
 
         return {}
 
+    def consensus_forward(
```
Comment (sjfleming) [inline cellarium/ml/models/nmf.py:589]: I recommend changing the name of this to `forward()`

File: cellarium/ml/models/nmf.py [hunk 6]
```diff
@@ -73,34 +371,68 @@ class NonNegativeMatrixFactorization(CellariumModel, PredictMixin):
...
+        self._alpha_tol = 0.005
+
+        self.n_nmf = r
+        self.density_threshold = density_threshold
+        self.local_neighborhood_size = local_neighborhood_size
+
+        if self.mode == 'nmf':
+            self.forward = self.nmf_forward
+        else:
+            self.forward = self.consensus_forward
```
Comment (sjfleming) [inline cellarium/ml/models/nmf.py:422]: Based on comments below, I recommend getting rid of this

File: examples/cli_workflow/nmf_heart_config.yaml [hunk 7]
```diff
@@ -0,0 +1,100 @@
+# lightning.pytorch==2.2.1
```
Comment (sjfleming) [inline examples/cli_workflow/nmf_heart_config.yaml:1]: Let's include only one config file, and let's have it be the config file that is used by the notebook tutorial for cNMF.

File: notebooks/nmf_human_heart_consensu_tutorial.ipynb [hunk 8]
```diff
@@ -0,0 +1,465 @@
+{
```
Comment (sjfleming) [inline notebooks/nmf_human_heart_consensu_tutorial.ipynb:1]: Let's include headers and explanations in this notebook.  Make it look clear and understandable.

**Review by sjfleming — CHANGES_REQUESTED** [review]
Nice work!  I left several comments.  Please either make changes or reply to each of them so that we can figure out where to make changes.

Also I set things up so that the github actions tests run automatically on this pull request.  Make sure that these tests are all passing.  If tests fail, find and fix the problems.  (I would first try these command line commands: `make format` and `make typecheck`.)  Currently the `nmf` branch is passing tests (https://github.com/cellarium-ai/cellarium-ml/pull/196).

File: cellarium/ml/models/nmf.py [hunk 9]
```diff
@@ -1,47 +1,180 @@
...
+    torch.manual_seed(0)
+    k_range = pipeline[-1].k_range
+
+    consensus_stat = {}
+    for k in k_range:
+
+        D_rkg = getattr(pipeline[-1], f"D_{k}_rkg")
+        consensus_output = consensus(D_rkg=D_rkg, k=k,
+                                     density_threshold=density_threshold,
+                                     local_neighborhood_size=density_threshold)
```
Comment (sjfleming) [inline cellarium/ml/models/nmf.py:120]: I think this should say 
```python
local_neighborhood_size= local_neighborhood_size
```

File: cellarium/ml/models/nmf.py [hunk 10]
```diff
@@ -1,47 +1,180 @@
...
-    raise NotImplementedError
 
+    k_range = pipeline[-1].k_range
+    pipeline[-1].get_rec_error=True
+    device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+    # get the anndata object
+    adatas = [dataset.dadc.adatas[i].adata for i in range(min(maximum_anndata_files_to_download,
+                                                              len(dataset.dadc.adatas)))]
+    adata = anndata.concat(adatas, axis=0, merge="same")
```
Comment (sjfleming) [inline cellarium/ml/models/nmf.py:54]: These lines of code to not appear to be used in what follows

File: cellarium/ml/models/nmf.py [hunk 11]
```diff
@@ -1,47 +1,180 @@
...
+        the_best_k: select the K to get cell embedding.
 
     Returns:
-        Tensor: single element tensor
+        AnnData with cell embeddings in adata.obsm[obsm_key_added]
     """
-    return torch.distributions.Poisson(rate=input).log_prob(target).sum()
 
+    pipeline[-1].get_rec_error=False
+    pipeline[-1].the_best_k=20
```
Comment (sjfleming) [inline cellarium/ml/models/nmf.py:86]: This seems to be overwriting the input `the_best_k`, which is being ignored

File: cellarium/ml/models/nmf.py (hunk 9 above)
Comment (ImXman) [reply cellarium/ml/models/nmf.py]: Corrected

File: cellarium/ml/cli.py (hunk 1 above)
Comment (sjfleming) [reply cellarium/ml/cli.py:489]: done

**Pull request merged**